package/MDAnalysis/converters/RDKitInferring.py

```diff
@@ -0,0 +1,703 @@
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+# doi: 10.25080/majora-629e541a-00e
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+"""RDKit bond order inferring --- :mod:`MDAnalysis.converters.RDKitInferring`
+=============================================================================
+Bond order and formal charge inferring for RDKit molecules.
+
+Classes
+-------
+
+.. autoclass:: MDAnalysisInferer
+   :members:
+   :private-members:
+
+.. autoclass:: TemplateInferer
+   :members:
+
+.. autoclass:: RDKitInferer
+   :members:
+
+.. autofunction:: sanitize_mol
+
+.. autofunction:: reorder_atoms
+
+"""
+import warnings
+from collections import defaultdict
+from contextlib import suppress
+from dataclasses import dataclass
+from typing import ClassVar, Dict, List, Optional, Tuple
+
+import numpy as np
+
+with suppress(ImportError):
+    from rdkit import Chem
+    from rdkit.Chem.AllChem import AssignBondOrdersFromTemplate
+    from rdkit.Chem.rdChemReactions import ChemicalReaction, ReactionFromSmarts
+
+    RDBONDORDER = {
+        1: Chem.BondType.SINGLE,
+        1.5: Chem.BondType.AROMATIC,
+        "ar": Chem.BondType.AROMATIC,
+        2: Chem.BondType.DOUBLE,
+        3: Chem.BondType.TRIPLE,
+    }
+    # add string version of the key for each bond
+    RDBONDORDER.update({str(key): value for key, value in RDBONDORDER.items()})
+    PERIODIC_TABLE = Chem.GetPeriodicTable()
+
+with suppress(ImportError):
+    from rdkit.Chem.rdDetermineBonds import (
+        DetermineBondOrders,
+    )  # available since 2022.09.1
+
+
+def reorder_atoms(
+    mol: "Chem.Mol", field: str = "_MDAnalysis_index"
+) -> "Chem.Mol":
+    """Reorder atoms based on the given field. Defaults to sorting in the same
+    order as the input AtomGroup.
+
+    Notes
+    -----
+    If the field is not found on the molecule, skips reordering.
+    """
+    with suppress(KeyError):
+        order = np.argsort([atom.GetIntProp(field) for atom in mol.GetAtoms()])
+        return Chem.RenumberAtoms(mol, order.astype(int).tolist())
+    # not a molecule converted by MDAnalysis
+    warnings.warn(
+        f"{field!r} not available on the input mol atoms, skipping reordering "
+        "of atoms."
+    )
+    return mol
+
+
+def sanitize_mol(mol: "Chem.Mol") -> None:
+    """Tries to sanitize the molecule, and if an error is raised, logs it as a
+    warning.
+    """
+    err = Chem.SanitizeMol(mol, catchErrors=True)
+    if err:
+        warnings.warn(
+            f"Could not sanitize molecule: failed during step {err!r}"
+        )
+
+
+@dataclass(frozen=True)
+class MDAnalysisInferer:
+    """Bond order and formal charge inferring as originally implemented for
+    the RDKit converter.
+
+    Attributes
```

Comment:
note to self: would be nice to have sanitization as a parameter rather than enforcing it. This could help users build their own inferring function on top of this one for cases that fail by simply adding a custom reaction to fix whichever part of their molecule is not inferred correctly.